File: apps/mobile/test/eta_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:servis_takip/data/eta/eta_provider.dart';
import 'package:servis_takip/data/eta/mock_eta_provider.dart';
import 'package:servis_takip/data/mock/demo_data.dart';

void main() {
  final provider = MockEtaProvider();

  test('hedef durak için kalan durak sayısını doğru hesaplar', () {
    final result = provider.calculate(EtaInput(
      vehicleLocation: demoStops.first.point,
      stops: demoStops,
      nextStopIndex: 1,
      targetStopIndex: 4,
      plannedArrivalAt: DateTime.now(),
    ));
    expect(result.remainingStops, 4);
    expect(result.remainingDistanceMeters, greaterThan(0));
    expect(result.etaMinutes, greaterThanOrEqualTo(0));
  });

  test('araç ilerledikçe ETA azalır', () {
    final far = provider.calculate(EtaInput(
      vehicleLocation: demoStops[0].point,
      stops: demoStops,
      nextStopIndex: 1,
      targetStopIndex: 4,
      plannedArrivalAt: DateTime.now(),
    ));
    final near = provider.calculate(EtaInput(
      vehicleLocation: demoStops[3].point,
      stops: demoStops,
      nextStopIndex: 4,
      targetStopIndex: 4,
      plannedArrivalAt: DateTime.now(),
    ));
    expect(near.remainingDistanceMeters, lessThan(far.remainingDistanceMeters));
    expect(near.remainingStops, lessThan(far.remainingStops));
  });
}

File: apps/mobile-legacy-rn/src/__tests__/eta.test.ts
import { MockETAProvider } from '@/services/eta/MockETAProvider';
import { demoStops } from '@/mocks/demoData';

describe('MockETAProvider', () => {
  const provider = new MockETAProvider();

  it('hedef durak için kalan durak sayısını doğru hesaplar', async () => {
    const result = await provider.calculate({
      vehicleLocation: { latitude: demoStops[0]!.latitude, longitude: demoStops[0]!.longitude },
      stops: demoStops,
      nextStopIndex: 1,
      targetStopIndex: 4,
      averageSpeedKmh: 32,
      dwellSecondsPerStop: 45,
      plannedArrivalAt: new Date().toISOString(),
    });
    // 1..4 arası = 4 durak
    expect(result.remaining_stops).toBe(4);
    expect(result.remaining_distance_meters).toBeGreaterThan(0);
    expect(result.eta_minutes).toBeGreaterThanOrEqual(0);
  });

  it('araç ilerledikçe ETA azalır', async () => {
    const base = {
      stops: demoStops,
      targetStopIndex: 4,
      averageSpeedKmh: 32,
      dwellSecondsPerStop: 45,
      plannedArrivalAt: new Date().toISOString(),
    };
    const far = await provider.calculate({
      ...base,
      nextStopIndex: 1,
      vehicleLocation: { latitude: demoStops[0]!.latitude, longitude: demoStops[0]!.longitude },
    });
    const near = await provider.calculate({
      ...base,
      nextStopIndex: 4,
      vehicleLocation: { latitude: demoStops[3]!.latitude, longitude: demoStops[3]!.longitude },
    });
    expect(near.remaining_distance_meters).toBeLessThan(far.remaining_distance_meters);
    expect(near.remaining_stops).toBeLessThan(far.remaining_stops);
  });
});
